torch/distributed/fsdp/optim_utils.py

```diff
@@ -346,10 +346,7 @@ def _flatten_tensor_optim_state(
             state).
     """
     non_none_tensors = [t for t in pos_dim_tensors if t is not None]
-    # Check that all are tensors on CPU with the same dtype
```

Comment:
Just removing this check, which is overly strict. We can just move the tensors to CPU inside the function to avoid device mismatch. It is not actually semantically important which device the tensors in the optimizer state are on.